resources/Microsoft.Windows.Setting.WindowsUpdate/Microsoft.Windows.Setting.WindowsUpdate.psm1

```diff
@@ -0,0 +1,377 @@
+
+    foreach ($key in $RegistryKeyProperty.Keys)
+    {
+        $value = $RegistryKeyProperty[$key]
+        $typeInfo = $value.GetType().Name
+
+        if ($typeInfo -eq 'Boolean')
+        {
+            $value = [int]$value
+        }
+
+        # validate the value of UserChoiceActiveHoursEnd and UserChoiceActiveHoursStart to be between 0 and 24
+        Assert-UserChoiceValue -KeyName $key -Value $value
+
+        # validate the value of DownloadRateBackgroundPct, DownloadRateForegroundPct and UpRatePctBandwith to be between 0 and 100
+        Assert-RatePercentageValue -KeyName $key -Value $value
+
+        # validate the value of UpRatePctBandwith to be between 5 and 500
+        Assert-UpRateValue -KeyName $key -Value $value
+
+        if (-not (DoesRegistryKeyPropertyExist -Path $Path -Name $key))
+        {
+            $null = New-ItemProperty -Path $Path -Name $key -Value $value -PropertyType 'DWord' -Force  
+        }
+
+        Write-Verbose -Message "Setting $key to $($RegistryKeyProperty[$key])"
+        Set-ItemProperty -Path $Path -Name $key -Value $value
+    }
+} 
+
+function Assert-UpRateValue 
+{
+    param (
+        [Parameter(Mandatory)]
+        [string] $KeyName,
+
+        [Parameter(Mandatory)]
+        [int] $Value
+    )
+
+    if ($KeyName -eq 'UpRatePctBandwidth' -and $Value -notin (5..500))
+    {
+        Throw "You are specifying a percentage value, which must be between 5 and 500. The value you provided is $Value. Please provide a value between 5 and 500."
+    }
+}
+
+function Assert-RatePercentageValue
+{
+    param (
+        [Parameter(Mandatory)]
+        [string] $KeyName,
+
+        [Parameter(Mandatory)]
+        [int] $Value
+    )
+
+    if ($KeyName -in ('DownloadRateBackgroundPct', 'DownloadRateForegroundPct', 'UpRatePctBandwidth') -and $Value -notin (0..100))
+    {
+        # TODO: It might be beneficial to add `Reasons` and not throw, only return statement
+        Throw "You are specifying a percentage value, which must be between 0 and 100. The value you provided is $Value. Please provide a value between 0 and 100."
+    }
+}
+
+function Assert-UserChoiceValue 
+{
+    param (
+        [Parameter(Mandatory)]
+        [string] $KeyName,
+
+        [Parameter(Mandatory)]
+        [int] $Value
+    )
+
+    if ($KeyName -in ('UserChoiceActiveHoursEnd', 'UserChoiceActiveHoursStart') -and $Value -notin (0..24))
+    {
+        Throw "Value must be between 0 and 24"
+    }
+}
+
+function Assert-DownloadRate
+{
+    param (
+        [Parameter(Mandatory)]
+        [hashtable] $Parameters
+    )
+
+    if ($Parameters.ContainsKey('DownloadRateBackgroundPct') -or $Parameters.ContainsKey('DownloadRateForegroundPct'))
+    {
+        if ($Parameters.ContainsKey('DownloadRateBackgroundBps') -or $Parameters.ContainsKey('DownloadRateForegroundBps'))
+        {
+            Throw "Cannot set both DownloadRateBackgroundPct/DownloadRateForegroundPct and DownloadRateBackgroundBps/DownloadRateForegroundBps"
```

Comment:
Instead of doing this, you could use the `ParameterSetName` to create two different parameter sets and let PowerShell do the validation

Reply:
I agree `ParameterSetName` would be great for functions. However, it doesn't work for classes (as I'm aware of). The functions are called through the class and there isn't really away to determine it in the class what the user has defined. That's why I'm validating it "later", rather than earlier ...

Reply:
According to the [documentation](https://learn.microsoft.com/en-us/powershell/dsc/concepts/class-based-resources?view=dsc-2.0#validation-property-attributes) any validation atteibute should work

Reply:
I might overlook things, but how is this going to work:

```powershell
[DscProperty()]
[Parameter(ParameterSetName = 'ByRate')]
[nullable[int]] $DownloadRateForegroundBps
```

Reply:
Now that is interesting . . . 

I checked out this PR locally and attempted to use the parameter set names myself. Even though the documentation *says* they should work, they absolutely do not. Normally it would throw an error that the sets are mutually exclusive, but it didn't, even when I double checked that my ErrorActionPreference was set to stop. 

I then tried to see if this could be done using a `ValidateScript` - but that gives the error that the attribute argument must be a constant. So either I'm doing something wrong, or not all of the validation attributes are actually supported on Class-based DSC Resources

Reply:
It was indeed. You nailed it by at least trying. I didn't put in the effort because, well... I guess in my head it didn't seem logical to me, nor have I seen it in other implementation.